addons/test_mail_full/tests/test_portal.py

```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+            if group_name == 'portal_customer' and group_func(self.customer):
+                return group_data['button_access']['url']
+        raise AssertionError('Record access URL not found')
+
+    def _get_composer_context(self, template_id=False):
+        return {
+            'default_model': self.record_customer._name,
+            'default_res_id': self.record_customer.id,
+            'default_use_template': template_id,
+            'default_template_id': template_id,
+            'default_composition_mode': 'comment',
+            'mark_so_as_sent': True,
+            'default_email_layout_xmlid': 'mail.mail_notification_layout_with_responsible_signature',
+            'force_email': True,
+            'model_description': 'Quotation',
+        }
+
+    @users('employee')
+    def test_send_message_to_customer_using_template(self):
+        """Send a mail to a customer without an account and check that it contains a link to view the record.
+
+        Other tests below check that that same link has the correct behavior.
+        This test follows the common use case by using a template while the next send the mail without a template."""
+        composer = self.env['mail.compose.message'].with_context(self._get_composer_context(self.mail_template.id))\
+            .create({})
+        # currently onchange necessary
+        composer._onchange_template_id_wrapper()
+
+        with self.mock_mail_gateway(mail_unlink_sent=True):
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 1)
+        self.assertIn(self._get_record_access_url(), self._mails[0].get('body').replace('&amp;', '&'))
+        self.assertEqual('Your quotation', self._mails[0].get('subject'))  # Check that the template is used
+
+    @users('employee')
+    def test_send_message_to_customer(self):
+        """Same as test_send_message_to_customer_using_template but without a template."""
+        composer = self.env['mail.compose.message'].with_context(self._get_composer_context()).create({
+            'subject': 'Quotation',
+            'body': '<p>Hello, your quotation is ready for review.</p>',
+            'partner_ids': self.customer.ids,
+        })
+
+        with self.mock_mail_gateway(mail_unlink_sent=True):
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 1)
+        self.assertIn(self._get_record_access_url(), self._mails[0].get('body').replace('&amp;', '&'))
+        self.assertEqual('Quotation', self._mails[0].get('subject'))  # Check that the template is not used
+
+    def test_customer_access_not_logged(self):
+        """Check that the access link redirects the customer (not logged) to the portal for viewing the record."""
+        res = self.url_open(self._get_record_access_url())
+        self.assertEqual(res.status_code, 404)  # No controller implemented for the test
```

Comment:
having a 404 error here because the route is not implemented is a bit mhe, can we implement the route ? (so in the test we will have a 200 code, which is more explicit)

Reply:
Yes that’s better to test the whole flow, thanks.